File: app/src/main/java/com/qbapps/claudeusage/data/local/UsageHistoryStore.kt
package com.qbapps.claudeusage.data.local

import android.content.Context
import com.qbapps.claudeusage.domain.model.ClaudeUsage
import com.qbapps.claudeusage.domain.model.UsageHistoryPoint
import dagger.hilt.android.qualifiers.ApplicationContext
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.SupervisorJob
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asStateFlow
import kotlinx.coroutines.launch
import kotlinx.coroutines.sync.Mutex
import kotlinx.coroutines.sync.withLock
import kotlinx.coroutines.withContext
import java.io.File
import java.time.Duration
import java.time.Instant
import javax.inject.Inject
import javax.inject.Singleton

/**
 * Persists usage history snapshots locally for trend visualizations.
 * Data is retained for [RETENTION_DAYS] and bounded to [MAX_POINTS] records.
 */
@Singleton
class UsageHistoryStore @Inject constructor(
    @ApplicationContext private val context: Context,
) {
    private val ioScope = CoroutineScope(SupervisorJob() + Dispatchers.IO)
    private val mutex = Mutex()
    private val historyFile: File by lazy { File(context.filesDir, HISTORY_FILE_NAME) }

    private val _history = MutableStateFlow<List<UsageHistoryPoint>>(emptyList())
    val history: StateFlow<List<UsageHistoryPoint>> = _history.asStateFlow()

    init {
        ioScope.launch {
            loadHistory()
        }
    }

    suspend fun append(usage: ClaudeUsage) = withContext(Dispatchers.IO) {
        val point = UsageHistoryPoint(
            timestamp = usage.fetchedAt,
            fiveHourUtilization = usage.fiveHour?.utilization,
            fiveHourResetsAt = usage.fiveHour?.resetsAt,
            sevenDayUtilization = usage.sevenDay?.utilization,
            sevenDayOpusUtilization = usage.sevenDayOpus?.utilization,
            sevenDaySonnetUtilization = usage.sevenDaySonnet?.utilization,
        )

        mutex.withLock {
            val updated = pruneAndTrim(_history.value + point)
            _history.value = updated
            writeHistory(updated)
        }
    }

    suspend fun clear() = withContext(Dispatchers.IO) {
        mutex.withLock {
            _history.value = emptyList()
            if (historyFile.exists()) {
                historyFile.delete()
            }
        }
    }

    private suspend fun loadHistory() = withContext(Dispatchers.IO) {
        mutex.withLock {
            if (!historyFile.exists()) {
                _history.value = emptyList()
                return@withLock
            }

            val parsed = historyFile.readLines()
                .mapNotNull(::parseLine)
                .sortedBy { it.timestamp }

            val pruned = pruneAndTrim(parsed)
            _history.value = pruned

            if (pruned.size != parsed.size) {
                writeHistory(pruned)
            }
        }
    }

    private fun writeHistory(points: List<UsageHistoryPoint>) {
        val content = buildString {
            points.forEach { point ->
                append(point.timestamp.toEpochMilli())
                append(',')
                append(point.fiveHourUtilization.orEmptyToken())
                append(',')
                append(point.fiveHourResetsAt.orEmptyEpochToken())
                append(',')
                append(point.sevenDayUtilization.orEmptyToken())
                append(',')
                append(point.sevenDayOpusUtilization.orEmptyToken())
                append(',')
                append(point.sevenDaySonnetUtilization.orEmptyToken())
                append('\n')
            }
        }
        historyFile.writeText(content)
    }

    private fun parseLine(line: String): UsageHistoryPoint? {
        val tokens = line.split(',', limit = 6)
        if (tokens.size !in 5..6) return null

        val timestampMs = tokens[0].toLongOrNull() ?: return null
        return if (tokens.size == 6) {
            UsageHistoryPoint(
                timestamp = Instant.ofEpochMilli(timestampMs),
                fiveHourUtilization = tokens[1].toDoubleOrNull(),
                fiveHourResetsAt = tokens[2].toLongOrNull()?.let { Instant.ofEpochMilli(it) },
                sevenDayUtilization = tokens[3].toDoubleOrNull(),
                sevenDayOpusUtilization = tokens[4].toDoubleOrNull(),
                sevenDaySonnetUtilization = tokens[5].toDoubleOrNull(),
            )
        } else {
            // Backward compatibility with older cache format (without resets_at).
            UsageHistoryPoint(
                timestamp = Instant.ofEpochMilli(timestampMs),
                fiveHourUtilization = tokens[1].toDoubleOrNull(),
                fiveHourResetsAt = null,
                sevenDayUtilization = tokens[2].toDoubleOrNull(),
                sevenDayOpusUtilization = tokens[3].toDoubleOrNull(),
                sevenDaySonnetUtilization = tokens[4].toDoubleOrNull(),
            )
        }
    }

    private fun pruneAndTrim(points: List<UsageHistoryPoint>): List<UsageHistoryPoint> {
        val cutoff = Instant.now().minus(Duration.ofDays(RETENTION_DAYS))
        val retained = points.filter { it.timestamp >= cutoff }
        return if (retained.size > MAX_POINTS) {
            retained.takeLast(MAX_POINTS)
        } else {
            retained
        }
    }

    private fun Double?.orEmptyToken(): String = this?.toString() ?: ""
    private fun Instant?.orEmptyEpochToken(): String = this?.toEpochMilli()?.toString() ?: ""

    companion object {
        private const val HISTORY_FILE_NAME = "usage_history.csv"
        private const val RETENTION_DAYS = 30L
        private const val MAX_POINTS = 10_000
    }
}

File: ios/OpenUsageCore/Sources/OpenUsageCore/Storage/UsageHistoryStore.swift
import Foundation

/// Faithful port of `data/local/UsageHistoryStore.kt`: a CSV file of raw usage
/// samples, retained for 30 days and capped at 10,000 rows. The on-disk format
/// is byte-compatible with the Android file, including the 5-column legacy rows.
public final class UsageHistoryStore: @unchecked Sendable {
    public static let retentionDays = 30
    public static let maxPoints = 10_000
    public static let fileName = "usage_history.csv"

    private let fileURL: URL
    private let lock = NSLock()
    private var cached: [UsageHistoryPoint]?

    public init(directory: URL) {
        self.fileURL = directory.appendingPathComponent(Self.fileName)
        try? FileManager.default.createDirectory(at: directory, withIntermediateDirectories: true)
    }

    /// Default location: `Application Support/OpenUsage`.
    public static func defaultDirectory() -> URL {
        let base = FileManager.default.urls(for: .applicationSupportDirectory, in: .userDomainMask).first
            ?? FileManager.default.temporaryDirectory
        return base.appendingPathComponent("OpenUsage", isDirectory: true)
    }

    public func history(now: Date = Date()) -> [UsageHistoryPoint] {
        lock.lock()
        defer { lock.unlock() }
        return loadLocked(now: now)
    }

    @discardableResult
    public func append(_ usage: ClaudeUsage, now: Date = Date()) -> [UsageHistoryPoint] {
        let point = UsageHistoryPoint(
            timestamp: usage.fetchedAt,
            fiveHourUtilization: usage.fiveHour?.utilization,
            fiveHourResetsAt: usage.fiveHour?.resetsAt,
            sevenDayUtilization: usage.sevenDay?.utilization,
            sevenDayOpusUtilization: usage.sevenDayOpus?.utilization,
            sevenDaySonnetUtilization: usage.sevenDaySonnet?.utilization
        )

        lock.lock()
        defer { lock.unlock() }
        let existing = loadLocked(now: now)
        let updated = Self.pruneAndTrim(existing + [point], now: now)
        cached = updated
        write(updated)
        return updated
    }

    public func clear() {
        lock.lock()
        defer { lock.unlock() }
        cached = []
        try? FileManager.default.removeItem(at: fileURL)
    }

    // MARK: - Private

    private func loadLocked(now: Date) -> [UsageHistoryPoint] {
        if let cached { return cached }

        guard let contents = try? String(contentsOf: fileURL, encoding: .utf8) else {
            cached = []
            return []
        }

        let parsed = contents
            .split(separator: "\n", omittingEmptySubsequences: true)
            .compactMap { Self.parseLine(String($0)) }
            .sorted { $0.timestamp < $1.timestamp }

        let pruned = Self.pruneAndTrim(parsed, now: now)
        cached = pruned
        if pruned.count != parsed.count { write(pruned) }
        return pruned
    }

    private func write(_ points: [UsageHistoryPoint]) {
        var content = ""
        for point in points {
            content += "\(epochMillis(point.timestamp)),"
            content += "\(Self.token(point.fiveHourUtilization)),"
            content += "\(Self.token(point.fiveHourResetsAt)),"
            content += "\(Self.token(point.sevenDayUtilization)),"
            content += "\(Self.token(point.sevenDayOpusUtilization)),"
            content += "\(Self.token(point.sevenDaySonnetUtilization))\n"
        }
        try? content.write(to: fileURL, atomically: true, encoding: .utf8)
    }

    static func token(_ value: Double?) -> String {
        guard let value else { return "" }
        return String(value)
    }

    static func token(_ value: Date?) -> String {
        guard let value else { return "" }
        return String(epochMillis(value))
    }

    static func parseLine(_ line: String) -> UsageHistoryPoint? {
        let tokens = line.split(separator: ",", omittingEmptySubsequences: false).map(String.init)
        guard tokens.count == 5 || tokens.count == 6 else { return nil }
        guard let timestampMs = Int64(tokens[0]) else { return nil }
        let timestamp = Date(timeIntervalSince1970: TimeInterval(timestampMs) / 1000.0)

        if tokens.count == 6 {
            return UsageHistoryPoint(
                timestamp: timestamp,
                fiveHourUtilization: Double(tokens[1]),
                fiveHourResetsAt: Int64(tokens[2]).map { Date(timeIntervalSince1970: TimeInterval($0) / 1000.0) },
                sevenDayUtilization: Double(tokens[3]),
                sevenDayOpusUtilization: Double(tokens[4]),
                sevenDaySonnetUtilization: Double(tokens[5])
            )
        }

        // Backward compatibility with the older cache format (no resets_at).
        return UsageHistoryPoint(
            timestamp: timestamp,
            fiveHourUtilization: Double(tokens[1]),
            fiveHourResetsAt: nil,
            sevenDayUtilization: Double(tokens[2]),
            sevenDayOpusUtilization: Double(tokens[3]),
            sevenDaySonnetUtilization: Double(tokens[4])
        )
    }

    static func pruneAndTrim(_ points: [UsageHistoryPoint], now: Date) -> [UsageHistoryPoint] {
        let cutoff = now.addingTimeInterval(-TimeInterval(retentionDays) * 86_400)
        let retained = points.filter { $0.timestamp >= cutoff }
        if retained.count > maxPoints {
            return Array(retained.suffix(maxPoints))
        }
        return retained
    }
}
